src/spikeinterface/core/waveform_tools.py

```diff
@@ -397,6 +398,280 @@ def _waveform_extractor_chunk(segment_index, start_frame, end_frame, worker_ctx)
+
+    s0, s1 = segment_slices[segment_index]
+    in_seg_spikes = spikes[s0:s1]
+
+    # take only spikes in range [start_frame, end_frame]
+    # this is a slice so no copy!!
+    i0 = np.searchsorted(in_seg_spikes["sample_index"], start_frame)
+    i1 = np.searchsorted(in_seg_spikes["sample_index"], end_frame)
+    if i0 != i1:
+        # protect from spikes on border :  spike_time<0 or spike_time>seg_size
+        # useful only when max_spikes_per_unit is not None
+        # waveform will not be extracted and a zeros will be left in the memmap file
+        while (in_seg_spikes[i0]["sample_index"] - nbefore) < 0 and (i0 != i1):
+            i0 = i0 + 1
+        while (in_seg_spikes[i1 - 1]["sample_index"] + nafter) > seg_size and (i0 != i1):
+            i1 = i1 - 1
+
+    # slice in absolut in spikes vector
+    l0 = i0 + s0
+    l1 = i1 + s0
+
+    if l1 > l0:
+        start = spikes[l0]["sample_index"] - nbefore
+        end = spikes[l1 - 1]["sample_index"] + nafter
+
+        # load trace in memory
+        traces = recording.get_traces(
+            start_frame=start, end_frame=end, segment_index=segment_index, return_scaled=return_scaled
+        )
+
+        for spike_ind in range(l0, l1):
+            sample_index = spikes[spike_ind]["sample_index"]
+            unit_index = spikes[spike_ind]["unit_index"]
+            wf = traces[sample_index - start - nbefore : sample_index - start + nafter, :]
+
+            if sparsity_mask is None:
+                all_waveforms[spike_ind, :, :] = wf
+            else:
+                mask = sparsity_mask[unit_index, :]
+                wf = wf[:, mask]
+                all_waveforms[spike_ind, :, : wf.shape[1]] = wf
+
+        if worker_ctx["mode"] == "memmap":
+            all_waveforms.flush()
+
+
+def split_waveforms_by_units(unit_ids, spikes, all_waveforms, sparsity_mask=None, folder=None):
+    """
+    Split a single buffer waveforms into waveforms by units (multi buffers or multi files).
+
+    Parameters
+    ----------
+    unit_ids: list or numpy array
+        List of unit ids
+    spikes: numpy array
+        The spike vector
+    all_waveforms : numpy array
+        Single buffer containing all waveforms
+    sparsity_mask : None or numpy array
+        Optionally the boolean sparsity mask
+    folder : None or str or Path
+        If a folde ri sgiven all
+
+    Returns
+    -------
+    waveforms_by_units: dict of array
+        A dict of arrays.
+        In case of folder not None, this contain the memmap of the files.
+    """
+    if folder is not None:
+        folder = Path(folder)
+    waveforms_by_units = {}
+    for unit_index, unit_id in enumerate(unit_ids):
+        mask = spikes["unit_index"] == unit_index
+        if sparsity_mask is not None:
+            chan_mask = sparsity_mask[unit_index, :]
+            num_chans = np.sum(chan_mask)
+            wfs = all_waveforms[mask, :, :][:, :, :num_chans]
+        else:
+            wfs = all_waveforms[mask, :, :]
```

Comment:
Why do you increase the dimensionality by one if mask is None?

Reply:
mask is the mask on first dim and this is not None this is line 657 not no increase normally.

Reply:
Right, thanks. I was confused with the sparsity mask that in most places in the code you call just mask. But this is the `spikes_in_the_unit` mask.